### PR TITLE
AR Setup with functionality 📱

### PR DESCRIPTION
**Pull request by orta**
Extends the UIVC from #2549 with the ability to keep track of whether you have given permissions for the camera.

### CODE REVIEW

File: Artsy.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1605,6 +1611,12 @@
...
 		6082B2041B46ADDE00953BC8 /* stubbed_image.png */ = {isa = PBXFileReference; lastKnownFileType = image.png; path = stubbed_image.png; sourceTree = "<group>"; };
+		6082C5AA2044A8D200F00635 /* Chevron_Black.png */ = {isa = PBXFileReference; lastKnownFileType = image.png; name = Chevron_Black.png; path = "Pods/Artsy+UILabels/Pod/Assets/Chevron_Black.png"; sourceTree = SOURCE_ROOT; };
+		6082C5AB2044A8D200F00635 /* Chevron_Black@2x.png */ = {isa = PBXFileReference; lastKnownFileType = image.png; name = "Chevron_Black@2x.png"; path = "Pods/Artsy+UILabels/Pod/Assets/Chevron_Black@2x.png"; sourceTree = SOURCE_ROOT; };
+		6082C5AE2044A8DF00F00635 /* Chevron_Black.png */ = {isa = PBXFileReference; lastKnownFileType = image.png; name = Chevron_Black.png; path = "Pods/Artsy+UILabels/Pod/Assets/Chevron_Black.png"; sourceTree = SOURCE_ROOT; };
+		6082C5AF2044A8DF00F00635 /* Chevron_Black@2x.png */ = {isa = PBXFileReference; lastKnownFileType = image.png; name = "Chevron_Black@2x.png"; path = "Pods/Artsy+UILabels/Pod/Assets/Chevron_Black@2x.png"; sourceTree = SOURCE_ROOT; };
```
Comment (orta) [inline Artsy.xcodeproj/project.pbxproj]: bit suspicious

File: Artsy.xcodeproj/xcshareddata/xcschemes/Artsy.xcscheme [hunk 2]
```diff
@@ -100,6 +101,7 @@
       buildConfiguration = "Debug"
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
+      language = ""
       launchStyle = "0"
       useCustomWorkingDirectory = "NO"
       ignoresPersistentStateOnLaunch = "NO"
```
Comment (orta) [inline Artsy.xcodeproj/xcshareddata/xcschemes/Artsy.xcscheme]: OK, I might make a single PR for  just this

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Developer Specific file shouldn’t be changed</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Needs testing on a Phone if change is non-trivial</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_travis">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Artsy.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -3878,41 +3882,44 @@
 		608B703017D4A10F0088A56C /* Assets */ = {
 			isa = PBXGroup;
 			children = (
```
Comment (orta) [inline Artsy.xcodeproj/project.pbxproj]: I hit the "re-order" button. And yeah, looks like everything is completely inconsistent 🍡

File: Artsy/Constants/ARDefaults.h [hunk 4]
```diff
@@ -34,10 +34,16 @@ extern NSString *const ARPushNotificationFollowArtist;
...
+#pragma mark admin features
 
 extern NSString *const AREmissionHeadVersionDefault;
 
+#pragma mark -
+#pragma mark user permissions
+
+extern NSString *const ARAugmentedRealityCameraAccessGiven;
+extern NSString *const ARAugmentedRealityHasSuccessfullyRan;
+extern NSString *const ARAugmentedRealityHasSeenSetup;
```
Comment (orta) [inline Artsy/Constants/ARDefaults.h]: Constants that can be used throughout the app, these get stored and are reset when you uninstall the app.

File: Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.h [hunk 5]
```diff
@@ -1,7 +1,14 @@
...
-- (instancetype)initWithMovieURL:(NSURL *)movieURL;
+- (instancetype)initWithMovieURL:(NSURL *)movieURL config:(ARAugmentedRealityConfig *)config;
 
+/// Is AR even supported?
++ (BOOL)canOpenARView;
+/// Have they already given access, and placed a work?
++ (BOOL)canSkipARSetup:(NSUserDefaults *)defaults;
+/// Request access
++ (void)validateAVAccess:(NSUserDefaults *)defaults callback:(void (^)(bool allowedAccess))closure;
 @end
```
Comment (orta) [inline Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.h]: There are two ways a function can be added to an object:

* `-` is a function that lives on an instance object
* `+` is a function that lives on the class

As a bunch of these functions don't need to know the internal state of the screen, then moving them to class functions makes it obvious to others that there are no side effects in using these functions.

File: Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m [hunk 6]
```diff
@@ -2,27 +2,53 @@
...
 @interface ARAugmentRealitySetupViewController ()
 @property (nonatomic, copy) NSURL *movieURL;
+@property (nonatomic, strong) NSUserDefaults *defaults;
+@property (nonatomic, strong) ARAugmentedRealityConfig *config;
+
+@property (nonatomic, weak) UILabel *subtitleLabel;
+@property (nonatomic, weak) UIButton *button;
+
+@property (nonatomic, assign) BOOL hasSentToSettings;
 @end
```
Comment (orta) [inline Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m]: We had to drastically increase the amount of variables kept in order to have the changing states for the UI.

File: Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m [hunk 7]
```diff
@@ -2,27 +2,53 @@
...
 @implementation ARAugmentRealitySetupViewController
 
-- (instancetype)initWithMovieURL:(NSURL *)movieURL
+- (instancetype)initWithMovieURL:(NSURL *)movieURL config:(ARAugmentedRealityConfig *)config
 {
     self = [super init];
     _movieURL = movieURL;
+    _config = config;
+    _defaults = [NSUserDefaults standardUserDefaults];
     return self;
```
Comment (orta) [inline Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m]: We need to pass the `ARAugmentedRealityConfig` to the AR screen later, so it was added to the initializer for the object.

File: Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m [hunk 8]
```diff
@@ -2,27 +2,53 @@
...
 {
     [super viewDidLoad];
 
+    BOOL firstTime = ![self.defaults boolForKey:ARAugmentedRealityHasSeenSetup];
+    BOOL hasGivenAccess = [self.defaults boolForKey:ARAugmentedRealityCameraAccessGiven];
+
+    NSString *buttonTitle = firstTime && !hasGivenAccess ? needsAccessButtonTitle : hasDeniedAccessButtonTitle ;
+    NSString *subtitleText = firstTime && !hasGivenAccess ? needsAccessSubtitle: hasDeniedAccessSubtitle;
+    SEL nextButtonSelector = hasGivenAccess ? @selector(sendToSettings) : @selector(next);
+
```
Comment (orta) [inline Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m]: The tests showcase this logic pretty reasonably later on in the PR

File: Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m [hunk 9]
```diff
@@ -78,5 +108,88 @@ - (void)viewDidLoad
...
+- (void)viewWillAppear:(BOOL)animated
+{
+    [super viewWillAppear:animated];
+    [self.defaults setBool:YES forKey:ARAugmentedRealityHasSeenSetup];
+
+    // Re-run the validation steps when they've come back from settings
+    if (self.hasSentToSettings) {
+        [self next];
+    }
+}
```
Comment (orta) [inline Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m]: When the screen has been presented we get a callback, that's where we declare that the setup screen has been seen. This callback will _also_ happen when you get back from the settings, so I added a check for it and make it run the same code that checks for the permissions

File: Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m [hunk 10]
```diff
@@ -78,5 +108,88 @@ - (void)viewDidLoad
...
+
++ (BOOL)canOpenARView
+{
+    if (@available(iOS 11.0, *)) {
+        return YES;
+    } else {
+        return NO;
+    }
+}
```
Comment (orta) [inline Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m]: Hrm, I should change this to 11.3

File: Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m [hunk 11]
```diff
@@ -78,5 +108,88 @@ - (void)viewDidLoad
...
+
++ (BOOL)canSkipARSetup:(NSUserDefaults *)defaults
+{
+    BOOL access = [defaults boolForKey:ARAugmentedRealityCameraAccessGiven];
+    BOOL used = [defaults boolForKey:ARAugmentedRealityHasSuccessfullyRan];
+    return access && used;
+}
```
Comment (orta) [inline Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m]: This is for the Artwork page to decide whether to show this page or just jump to AR

File: Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m [hunk 12]
```diff
@@ -78,5 +108,88 @@ - (void)viewDidLoad
...
+    if ([defaults boolForKey:ARAugmentedRealityCameraAccessGiven]) {
+        return closure(YES);
+    }
+
+    [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [defaults setBool:granted forKey:ARAugmentedRealityCameraAccessGiven];
+            closure(granted);
+        });
+    }];
```
Comment (orta) [inline Artsy/View_Controllers/ARVIR/ARAugmentRealitySetupViewController.m]: `AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo` is the only way to check for permissions for AR. The interesting thing is that it uses the text as Consignments do.

screen shot 2018-02-28 at 12 41 43 pm

So I'll take a second stab at this later by directly trying to load AR when you press the button, which should show different permissions

File: Artsy_Tests/View_Controller_Tests/ARVIR/ARAugmentRealitySetupViewControllerSpec.m [hunk 13]
```diff
@@ -1,11 +1,92 @@
+@import Forgeries;
+
+#import "ARDefaults.h"
 #import "ARAugmentRealitySetupViewController.h"
 
+@interface ARAugmentRealitySetupViewController ()
+@property (nonatomic, copy) NSUserDefaults *defaults;
+
+- (void)back;
+@end
```
Comment (orta) [inline Artsy_Tests/View_Controller_Tests/ARVIR/ARAugmentRealitySetupViewControllerSpec.m]: The official interface for `ARAugmentRealitySetupViewController` does not mention that it has an `@property` for defaults, or a back function. However, I want to test them, so I can declare in the test that it exists and the compiler is gonna be fine with it.

File: Artsy_Tests/View_Controller_Tests/ARVIR/ARAugmentRealitySetupViewControllerSpec.m [hunk 14]
```diff
@@ -1,11 +1,92 @@
...
+    ARAugmentedRealityCameraAccessGiven: @(NO),
+    ARAugmentedRealityHasSeenSetup: @(YES),
+    ARAugmentedRealityHasSuccessfullyRan: @(YES)
+}];
+
+ForgeriesUserDefaults *completedDefaults = [ForgeriesUserDefaults defaults:@{
+    ARAugmentedRealityCameraAccessGiven: @(YES),
+    ARAugmentedRealityHasSeenSetup: @(YES),
+    ARAugmentedRealityHasSuccessfullyRan: @(YES)
+}];
```
Comment (orta) [inline Artsy_Tests/View_Controller_Tests/ARVIR/ARAugmentRealitySetupViewControllerSpec.m]: Sets up some example states a user could be in for testing, check out the awesome artwork for this Pod: https://cocoapods.org/pods/Forgeries

File: Artsy_Tests/View_Controller_Tests/Web_Browsing/ARExternalWebBrowserViewControllerTests.m [hunk 15]
```diff
@@ -21,7 +21,7 @@ @interface ARExternalWebBrowserViewController (Tests) <UIScrollViewDelegate>
 
 it(@"sets the scroll view's `delegate` if it belongs to ARNavigationController", ^{
     [vc ar_presentWithFrame:[UIScreen mainScreen].bounds];
-    ARNavigationController *nav =[[ARNavigationController alloc] initWithRootViewController:vc];
+    [[ARNavigationController alloc] initWithRootViewController:vc];
```
Comment (orta) [inline Artsy_Tests/View_Controller_Tests/Web_Browsing/ARExternalWebBrowserViewControllerTests.m]: I was geting a warning about this, better to kill those as I see them, doesn't change behavior anywhere

File: CHANGELOG.yml [hunk 16]
```diff
@@ -5,7 +5,7 @@ upcoming:
       - Fixes universal links in iOS 11 - sarah
 
       - ARVIR:
-        - Adds a new VIR intro screen - orta
+        - Adds a new VIR intro screen, which handles disabled states - orta
```
Comment (orta) [inline CHANGELOG.yml]: We don't have to add new CHANGELOG entries, just change the changelog 👩‍🚀. There's no point in adding redundant messages around the same topic.

**Pull request merged**